jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/CompactPathRule.java

```diff
@@ -14,28 +14,148 @@
 import org.eclipse.jetty.util.URIUtil;
 
 /**
- * <p>Rewrites the URI by compacting to remove occurrences of {@code //}.</p>
- * <p>For example, {@code //foo/bar//baz} is compacted to {@code /foo/bar/baz}.</p>
+ * <p>Rewrites the URI by compacting to remove occurrences of {@code //} and {@code /../}.</p>
+ *
+ * <p>Optionally decode the path before compacting.</p>
+ * <p>Optionally canonicalize path before compacting.</p>
+ *
+ * <table>
+ *     <caption>Examples</caption>
+ *     <tr>
+ *       <th>isDecoding</th>
+ *       <th>isCanonicalizing</th>
+ *       <th>Input Path</th>
+ *       <th>Resulting Path</th>
+ *     </tr>
+ *     <tr>
+ *       <td>false (default)</td>
+ *       <td>false (default)</td>
+ *       <td>{@code //foo/bar//baz}</td>
+ *       <td>{@code /foo/bar/baz}</td>
+ *     </tr>
+ *     <tr>
+ *       <td>false (default)</td>
+ *       <td>false (default)</td>
+ *       <td>{@code //foo/../bar//baz}</td>
+ *       <td>{@code /bar/baz}</td>
+ *     </tr>
+ *     <tr>
+ *       <td>true</td>
+ *       <td>true</td>
+ *       <td>{@code //foo/../bar//baz}</td>
+ *       <td>{@code /bar/baz}</td>
+ *     </tr>
+ * </table>
+ *
+ * @see URIUtil#decodePath(String)
+ * @see URIUtil#canonicalPath(String)
+ * @see URIUtil#compactPath(String)
  */
 public class CompactPathRule extends Rule
 {
+    public record CompactedEvent(Handler input, String changedPath, String message, CompactPathRule rule)
+    {
+    }
+
+    public interface Listener
+    {
+        void onCompactedEvent(CompactedEvent event);
+    }
+
+    private List<Listener> listeners = new ArrayList<>();
+    private boolean decoding = false;
+    private boolean canonicalizing = false;
+
+    public void addListener(Listener listener)
+    {
+        listeners.add(listener);
+    }
+
+    public boolean removeListener(Listener listener)
+    {
+        return listeners.remove(listener);
+    }
+
+    public boolean isCanonicalizing()
+    {
+        return canonicalizing;
+    }
+
+    public void setCanonicalizing(boolean canonicalizing)
+    {
+        this.canonicalizing = canonicalizing;
+    }
+
+    public boolean isDecoding()
+    {
+        return decoding;
+    }
+
+    public void setDecoding(boolean decoding)
+    {
+        this.decoding = decoding;
+    }
+
     @Override
     public Handler matchAndApply(Handler input) throws IOException
     {
+        // Get the path without extra things like path parameters
         String path = input.getHttpURI().getCanonicalPath();
-        String compacted = URIUtil.compactPath(path);
+        String cleanedPath = path;
 
-        if (path.equals(compacted))
-            return null;
+        if (isDecoding())
+            cleanedPath = URIUtil.decodePath(cleanedPath);
+        if (isCanonicalizing())
+        {
+            cleanedPath = URIUtil.canonicalPath(cleanedPath);
+            if (cleanedPath == null)
+            {
+                notifyPathCompaction(input, cleanedPath, "Navigate above root URL");
```

Comment:
naming a bit strange here.   I guess "canonicalization" is a "compaction"??
But calling this a compliance notification would be better.. if we have applicable compliance violations for all operations here.